source/grid/tria.cc

```diff
@@ -14315,7 +14306,7 @@ Triangulation<dim, spacedim>::prepare_coarsening_and_refinement()
           // refinement flags, but we will also have to remove
           // coarsening flags on cells adjacent to vertices that will
           // see refinement
-          for (cell = last_active(); cell != endc; --cell)
+          for (active_cell_iterator cell = last_active(); cell != end(); --cell)
```

Comment:
I guess it's not performance-critical to have a non-fixed loop bound here

Reply:
We had a situation a while back where evaluating ``end()`` of a triangulation inside the for loop condition lead to terrible O(n^2) performance. I think it might be worth caching ``end()`` in every situation.

Reply:
I had thought about this situation, but determined that it is not a problem here.

That situation you reference was different. It was a call to `end(level)`, which actually does computations. `end()` is implemented as
```
template <int dim, int spacedim>
typename Triangulation<dim, spacedim>::cell_iterator
Triangulation<dim, spacedim>::end() const
{
  return cell_iterator(const_cast<Triangulation<dim, spacedim> *>(this),
                       -1,
                       -1);
}
```
in the same file (so visible to the compiler), which is easily optimized away.

I should try to actually mark this function as `constexpr`, but would like to leave that for a separate PR.

Reply:
Marking the function as `constexpr` will not have any effect unless assigning to a `constexpr` variable (https://godbolt.org/z/FPMvvn).

Reply:
But what I'm trying to say is that if the function is marked `constexpr`, then it's clear to the compiler that its result can be hoisted out of the loop. But that's the case nonetheless here.

Reply:
I just tried what I would need to make `end()`  `constexpr` and it really hard to do this in a `C++11`-conforming way. You would need mark a lot more constructors `constexpr`. The most problematic functions that need to be marked `constexpr` are the constructor for `TriaAccessorBase` and the constructor for `TriaIterator`.